iothub/device/tests/Mqtt/MqttIotHubAdapterTests.cs

```diff
@@ -0,0 +1,99 @@
+namespace Microsoft.Azure.Devices.Client.Test.Mqtt
+{
+    using DotNetty.Codecs.Mqtt.Packets;
+    using DotNetty.Transport.Channels;
+    using Microsoft.Azure.Devices.Client.Transport.Mqtt;
+    using Microsoft.VisualStudio.TestTools.UnitTesting;
+    using Moq;
+    using System;
+    using System.Threading.Tasks;
+
+    [TestClass]
+    [TestCategory("Unit")]
+    public class MqttIotHubAdapterTest
+    {
+        [TestMethod]
+        public async Task ShouldNotAssumeEventOrdering_SubscribeContinuesBeforeSubAckReceived()
+        {
+            var request = new SubscribePacket(0, new SubscriptionRequest("myTopic", QualityOfService.AtMostOnce));
+            Func<PacketWithId, SubAckPacket> ackFactory =
+                (packet) =>
+                {
+                    return new SubAckPacket() { PacketId = packet.PacketId };
+                };
+
+            await SendRequestAndAcknowledgementsInSpecificOrder(request, ackFactory, false).ConfigureAwait(false);
```

Comment:
Nit: last attribute should be named for clarity (receiveResponseBeforeSendingRequestContinues: false).